rasterio/vrt.py

```diff
@@ -245,3 +245,57 @@ def _boundless_vrt_doc(
         dstrect.attrib['ySize'] = str(src_dataset.height)
 
     return ET.tostring(vrtdataset, encoding="utf-8")
+
+
+class VirtualDataset:
+    """A lazily evaluated virtual dataset based on GDAL's VRT."""
+
+    @classmethod
+    def fromstring(cls, text):
+        """Create a virtual dataset from XML.
+
+        Paramters
+        ---------
+        text : bytes or str
+            XML as a byte or unicode string.
+
+        Returns
+        -------
+        VirtualDataset
+
+        """
+        # To avoid a circular import.
+        from rasterio.io import MemoryFile
+
+        tree = ET.fromstring(text)
+        doc = ET.tostring(tree)
+        memfile = MemoryFile(doc, ext=".vrt")
```

Comment:
Here's the temporary file we need. VirtualDataset wraps an XML string and uses a MemoryFile under the hood.

Reply:
This might be a possible issue that needs to be looked at. When passing invalid xml to MemoryFile, I get the following traceback. The closed attribute of MemoryFile reports that it is not closed. It is unclear if the MemoryFile is getting completely cleaned up.
```
---------------------------------------------------------------------------
CPLE_AppDefinedError                      Traceback (most recent call last)
Cell In [62], line 1
----> 1 z.close()

File rasterio/_io.pyx:1229, in rasterio._io.MemoryFileBase.close()

File rasterio/_io.pyx:275, in rasterio._io._delete_dataset_if_exists()

File rasterio/_io.pyx:276, in rasterio._io._delete_dataset_if_exists()

File rasterio/_base.pyx:218, in rasterio._base.open_dataset()

File rasterio/_err.pyx:221, in rasterio._err.exc_wrap_pointer()

CPLE_AppDefinedError: Parse error at EOF, not all elements have been closed, starting with SimpleSource
```

Reply:
@groutr thanks for pointing this out! It revealed a lower level bug that I'm about to fix.